pkg/filesystem/subvolume.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2023 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subvolume
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type fsStruct struct {
+	Name string
+	data []string
+}
+
+type subVolumeInfo struct {
+	svg string
+	fs  string
+}
+
+func List(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool) {
+
+	subvolumeNames := getK8sRefSubvolume(ctx, clientsets)
+	listCephFSSubvolumes(ctx, clientsets, operatorNamespace, clusterNamespace, includeStale, subvolumeNames)
+
+}
+
+// getk8sRefSubvolume returns the k8s ref for the subvolumes
+func getK8sRefSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets) map[string]subVolumeInfo {
+	pvList, err := clientsets.Kube.CoreV1().PersistentVolumes().List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("Error fetching PVs: %v\n", err))
+	}
+	subvolumeNames := make(map[string]subVolumeInfo)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI.VolumeAttributes["subvolumeName"] != "" {
+			subvolumeNames[pv.Spec.CSI.VolumeAttributes["subvolumeName"]] = subVolumeInfo{}
+		}
+	}
+	return subvolumeNames
+}
+
+// listCephFSSubvolumes list all the subvolumes
+func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string, includeStale bool, subvolumeNames map[string]subVolumeInfo) map[string]subVolumeInfo {
+
+	// getFilesystem gets the filesystem
+	fsstruct := getFileSystem(ctx, clientsets, operatorNamespace, clusterNamespace)
+	staleSubvolumeNames := make(map[string]subVolumeInfo)
+	var svList string
+	if includeStale {
+		fmt.Println("Filesystem  Subvolume  SubvolumeGroup")
```

Comment:
We always print the "Stale" column, right? No need for this?
```suggestion
```

Reply:
printing stale cloumn here won't make sense, as we already pass on the --stale flag. if we do this, then there is no use of adding flag to ls. So IMHO, this is fine.